stellargraph/data/unsupervised_sampler.py

```diff
@@ -121,69 +123,52 @@ def generator(self, batch_size):
                 This must be an even number.
 
         Returns:
-            Tuple of lists of target/context pairs and labels – 0 for a negative and 1 for a
-             positive pair: ([[target, context] ,... ], [label, ...])
+            List of batches, where each batch is a tuple of (list context pairs, list of labels)
         """
         self._check_parameter_values(batch_size)
 
-        positive_pairs = list()
-        negative_pairs = list()
-
-        sample_counter = 0
-
         all_nodes = list(self.graph.nodes())
-
         # Use the sampling distribution as per node2vec
         degrees = self.graph.node_degrees()
         sampling_distribution = [degrees[n] ** 0.75 for n in all_nodes]
 
-        done = False
-        while not done:
-            self.random.shuffle(self.nodes)
-            for node in self.nodes:  # iterate over root nodes
-                # Get 1 walk at a time. For now its assumed that its a uniform random walker
-                walk = self.walker.run(
-                    nodes=[node],  # root nodes
-                    length=self.length,  # maximum length of a random walk
-                    n=1,  # number of random walks per root node
+        walks = self.walker.run(
+            nodes=self.nodes, length=self.length, n=self.number_of_walks
+        )
+
+        # first item in each walk is the target/head node
+        targets = [walk[0] for walk in walks]
+
+        positive_pairs = np.array(
+            [
+                (target, positive_context)
+                for target, walk in zip(targets, walks)
+                for positive_context in walk[1:]
+            ]
+        )
+
+        negative_samples = self.random.choices(
+            all_nodes, weights=sampling_distribution, k=len(positive_pairs)
+        )
+        negative_pairs = np.array(
+            [
+                (target, negative_context)
+                for (target, _), negative_context in zip(
+                    positive_pairs, negative_samples
                 )
+            ]
+        )
+
+        # shuffle
+        pairs = np.concatenate((positive_pairs, negative_pairs), axis=0)
+        labels = np.repeat([1, 0], len(positive_pairs))
+
+        indices = self.np_random.permutation(len(pairs))
```

Comment:
I think this very slightly changes behaviour: previously, I think there was always an equal number of positive and negative examples in each batch, but now they're shuffled together, arbitrarily, and they may be uneven (although, equal on average).

Is that a problem?

Reply:
That sounds reasonable to me, I wonder if it's worth having a brief comment?

```suggestion
        # don't ensure an equal number of positive/negative examples in each batch,
        # just an equal number overall
        indices = self.np_random.permutation(len(pairs))
```

or if that's pointless?